src/commands/watch.rs

```diff
@@ -78,182 +73,177 @@ pub struct WatchCommand {
+        // Start the the uppificator and buildifier to process notifications from the
+        // `watchexec` tasks.
 
-        // We always want to watch the application manifest
-        let manifest_pattern = WatchPattern::new(
-            app.to_str()
-                .with_context(|| format!("non-unicode manifest path {:?}", app))?
-                .to_owned(),
-            app_dir.as_path(),
-        )?;
-
-        // We want to watch the source code if we aren't skipping the build step
-        let source_patterns = match !self.skip_build {
-            false => vec![],
-            true => app_manifest
-                .components
-                .iter()
-                .filter_map(|c| WatchCommand::create_source_pattern(c, &app_dir))
-                .flatten()
-                .collect::<Result<Vec<WatchPattern>>>()?,
+        let uppificator_handle = tokio::task::spawn(async move { uppificator.run().await });
+        let buildifier_handle = if self.skip_build {
+            tokio::task::spawn(tokio::time::sleep(tokio::time::Duration::MAX))
+        } else {
+            tokio::task::spawn(async move { buildifier.run().await })
         };
 
-        // We want to watch a component's source if it has no build step. If we're skipping
-        // building then we'll watch all component sources.
-        let component_source_patterns = app_manifest
+        // Wait for either the user to stop the watch, or some catastrophe to dump us out.
+
+        _ = ctrlc::set_handler(move || {
+            _ = stop_tx.send(Uuid::new_v4());
+        });
+
+        _ = futures::future::select_all(vec![
+            uppificator_handle,
+            buildifier_handle,
+            artifact_watcher_handle,
+            build_files_watcher_handle,
+        ])
+        .await;
+
+        Ok(())
+    }
+
+    fn spawn_watchexec(
+        &self,
+        manifest_dir: &Path,
+        filterer: impl watchexec::filter::Filterer + 'static,
+        notifier: tokio::sync::watch::Sender<Uuid>,
+        impact_description: &'static str,
+    ) -> anyhow::Result<tokio::task::JoinHandle<()>> {
+        let mut rt = watchexec::config::RuntimeConfig::default();
+        rt.pathset([manifest_dir]);
+        rt.filterer(Arc::new(filterer));
+        rt.action_throttle(Duration::from_millis(self.debounce));
+        rt.on_action(move |action: watchexec::action::Action| {
+            tracing::info!(
+                "spin watch detected changes requiring {impact_description}: {}",
+                paths_of(&action)
+            );
+            _ = notifier.send(Uuid::new_v4());
+            action.outcome(watchexec::action::Outcome::DoNothing);
+            async { Ok::<(), Infallible>(()) }
+        });
+        let watcher = Watchexec::new(watchexec::config::InitConfig::default(), rt)?;
+        let watcher_jh = tokio::task::spawn(async move {
+            _ = watcher.main().await;
+        });
+        Ok(watcher_jh)
+    }
+
+    async fn build_filterer<T>(
```

Comment:
Yes the existing ignore stuff is definitely something I skipped while trying to get this working, and wanted to understand from you.  My impression was that the existing implementation has only one ignore path, `*.swp`, and with the dual-watchexec thing it would be applied to both filters, correct?  I'm not sure what this pattern is though, so if you have time to walk me through it, that would be super useful - thanks!

Reply:
You are correct that the original implementation only had the one ignore path. I originally planned to not support ignore paths but ran into issues when I was testing with Vim where the swap files it would create would cause a deluge  of restarts. My very hack solution was to introduce this very naive ignore list to the filter with only the single path for now. It's hack because we don't expose this functionality to the users at all and because the pattern doesn't actually cover all the possible swap files Vim could make (see [here](https://unix.stackexchange.com/questions/326707/vim-what-are-all-the-possible-swapfile-extensions)).

To maintain backwards compatibility I think we should support ignoring any files matching `*.swp`. Going forward I think it would be good to expose this through the `spin.toml` configuration so users can get more granular about what they want to ignore. Perhaps through gitignore syntax like `!*.swp`.

As to whether this should be applied to both filters — that is a good question. My gut reaction is that yes it should be applied to both. For example if you were editing a static asset with Vim it would create a swap file that would match the static assets. Imagining the future where we support user provided ignore paths it is less clear to me if they should be added to both filters or not. That's maybe an issue we can address at that point.

If you need more explanation ping me and we can chat about it.